rfcs/20200624-pluggable-device-for-tensorflow.md

```diff
@@ -0,0 +1,251 @@
+
+<div align=center> 
+<img src=20200624-pluggable-device-for-tensorflow/design_overview.png>
+</div>
+
+* `PluggableDevice` is defined in TensorFlow proper which inherits from [LocalDevice](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/local_device.h).It is built on top of  StreamExecutor C++ interface to manage `PluggableDevice`’s key abstractions like StreamExecutor, stream, memory and event.
+
+* `PluggableDeviceExecutor` implements [StreamExecutor](https://github.com/tensorflow/tensorflow/blob/e5023a1738cce7efcdf9d87863b85c80ab2f8c9e/tensorflow/stream_executor/stream_executor_pimpl.h#L73) and is built on top of StreamExecutor C API (addressed in [RFC](https://github.com/tensorflow/community/pull/257)). 
+
+* `PluggableDevice Implementation` is inside the TensorFlow plugin, which provides those C functions implementation defined in the StreamExecutor C API.
+
+The pluggable device mechanism contains device discovery and creation process which creates a `PluggableDevice` object and `PluggableDeviceExecutor` object for each pluggable device. 
+
+With the RFC, existing TensorFlow GPU programs can run on a plugged device without the user changing the code. The Diagram 2 describes the workflow of TensorFlow with device plugin, it shows how a simple GPU program runs on the pluggable device.
+<div align="center">
+<img src=20200624-pluggable-device-for-tensorflow/gpu_example.png>
+</div>
+
+### Device Discovery
+
+Upon initialization of TensorFlow, it uses platform independent `LoadLibrary()` to load the dynamic library. The plugin library should be installed to default plugin directory "…python_dir.../site-packages/tensorflow-plugins". The modular tensorflow [RFC](https://github.com/tensorflow/community/pull/77) describes the process of loading plugins. 
+
+During the plugin library initialization, it calls the `SE_ReigsterPlatform()` API to register the stream executor platform (`SE_Platform` struct) to TensorFlow proper. The `SE_ReigsterPlatform()` API is a callback API, part of StreamExecutor C API, which passes necessary information to TensorFlow proper to instantiate a stream executor platform ([se::platform](https://github.com/tensorflow/tensorflow/blob/cb32cf0f0160d1f582787119d0480de3ba8b9b53/tensorflow/stream_executor/platform.h#L93) class) and register to a global object [se::MultiPlatformManager](https://github.com/tensorflow/tensorflow/blob/cb32cf0f0160d1f582787119d0480de3ba8b9b53/tensorflow/stream_executor/multi_platform_manager.h#L82). 
+The stream executor platform must be registered with the name "PluggableDevice".  
```

Comment:
Why is the name load-bearing?

Reply:
Thanks for the reminder! This limitation can be removed if StreamExecutor C API can provide a  methodology to register the platform name to the PluggableDevice since PluggableDevice needs to find the corresponding StreamExecutor Platform through its name(`se::MultiPlatformManager::PlatformWithName`) during the creation. 
here is @annarev 's suggestion on the StreamExecutor C API implementation which may solve this issue:PluggableDeviceFactory takes the platform name.

void SE_RegisterPlatform(
    char* name, size_t name_len, char* type, size_t type_len, SE_Platform* platform, TF_Status* status) {
   // Register new platform
   ....
   // Register PluggableDevice
   std::string platform_name_str(name, name_len);
   std::string type_str(type, type_len);
   DeviceFactory::Register(type_str, new PluggableDeviceFactory(platform_name_str), priority);
}

Reply:
Yep, sorry for delay updated SE API doc. I will look at updating it now.